### ISSUE TITLE
dygraph 下实现CRNN+CTC的问题

### ISSUE THREAD

**Issue by wwjjy** [PaddlePaddle/models#3952]
您好，我想要使用动态图来实现CRNN+CTC，目前参考https://github.com/PaddlePaddle/models/tree/a92e212932b764e500a833527e0fb772ac9a491a/dygraph/ocr_recognition 这个项目实现了CRNN部分，但是该部分的输出是一个3-D的定长tensor，我看了fluid.layers.warpctc op如果使用定长tensor作为输入的话需要指定每个序列的长度，如果不指定的话直接运行会报 段错误，请问这边我需要怎么将CRNN部分的输出以及label转为lod_tensor嘞

**Comment by wwjjy** [comment]
尝试加入了input_length 和label_length两个参数，具体调用如下：
loss = fluid.layers.warpctc(input=out, label=label, input_length=pred_len, label_length=label_len, blank=train_parameters['class_dim'])

各个入参的形状如下：
out： [32, 128, 64]
label： [4096, 1]
pred_len： [128]
label_length： [128]
从形状上看符合预期，但是运行还是报错，错误如下：
W1120 15:19:40.449357   283 init.cc:212] *** Aborted at 1574234380 (unix time) try "date -d @1574234380" if you are using GNU date ***
W1120 15:19:40.451268   283 init.cc:212] PC: @                0x0 (unknown)
W1120 15:19:40.451530   283 init.cc:212] *** SIGSEGV (@0x0) received by PID 283 (TID 0x7f4babc04700) from PID 0; stack trace: ***
W1120 15:19:40.453199   283 init.cc:212]     @     0x7f4bab7e1390 (unknown)
W1120 15:19:40.455535   283 init.cc:212]     @     0x7f4b70d50f3d paddle::operators::WarpCTCKernel<>::Compute()
W1120 15:19:40.457269   283 init.cc:212]     @     0x7f4b70d53593 _ZNSt17_Function_handlerIFvRKN6paddle9framework16ExecutionContextEEZNKS1_24OpKernelRegistrarFunctorINS0_8platform9CUDAPlaceELb0ELm0EJNS0_9operators13WarpCTCKernelINS7_17CUDADeviceContextEfEEEEclEPKcSF_iEUlS4_E_E9_M_invokeERKSt9_Any_dataS4_
W1120 15:19:40.461097   283 init.cc:212]     @     0x7f4b731e7d74 paddle::imperative::PreparedOp::Run()
W1120 15:19:40.463454   283 init.cc:212]     @     0x7f4b70cb3817 paddle::imperative::OpBase::Run()
W1120 15:19:40.465337   283 init.cc:212]     @     0x7f4b70cbbbe8 paddle::imperative::Tracer::TraceOp()
W1120 15:19:40.465597   283 init.cc:212]     @     0x7f4b70b8bf10 _ZZN8pybind1112cpp_function10initializeIZN6paddle6pybind14BindImperativeEPNS_6moduleEEUlRNS2_10imperative6TracerERKSsRKSt13unordered_mapISsNS_6handleESt4hashISsESt8equal_toISsESaISt4pairIS9_SC_EEESM_SB_ISsN5boost7variantINSN_5blankEifSsSt6vectorIiSaIiEESQ_IfSaIfEESQ_ISsSaISsEEbSQ_IbSaIbEEPNS2_9framework9BlockDescElSQ_IS11_SaIS11_EESQ_IlSaIlEENSN_6detail7variant5void_ES18_S18_S18_S18_S18_S18_EESE_SG_SaISH_IS9_S19_EEERKNS2_8platform9CUDAPlaceEbE15_vIS8_SA_SM_SM_S1C_S1G_bEINS_4nameENS_9is_methodENS_7siblingEEEEvOT_PFT0_DpT1_EDpRKT2_ENUlRNS_6detail13function_callEE1_4_FUNES1Y_
W1120 15:19:40.467007   283 init.cc:212]     @     0x7f4b70ac1ed6 pybind11::cpp_function::dispatcher()
W1120 15:19:40.467540   283 init.cc:212]     @     0x5559b322e744 _PyMethodDef_RawFastCallKeywords
W1120 15:19:40.467990   283 init.cc:212]     @     0x5559b322e861 _PyCFunction_FastCallKeywords
W1120 15:19:40.468425   283 init.cc:212]     @     0x5559b329a6e8 _PyEval_EvalFrameDefault
W1120 15:19:40.468823   283 init.cc:212]     @     0x5559b31de539 _PyEval_EvalCodeWithName
W1120 15:19:40.469205   283 init.cc:212]     @     0x5559b322def5 _PyFunction_FastCallKeywords
W1120 15:19:40.469636   283 init.cc:212]     @     0x5559b3295a93 _PyEval_EvalFrameDefault
W1120 15:19:40.470041   283 init.cc:212]     @     0x5559b31de539 _PyEval_EvalCodeWithName
W1120 15:19:40.470445   283 init.cc:212]     @     0x5559b31df860 _PyFunction_FastCallDict
W1120 15:19:40.470836   283 init.cc:212]     @     0x5559b31fde53 _PyObject_Call_Prepend
W1120 15:19:40.471280   283 init.cc:212]     @     0x5559b31f0dbe PyObject_Call
W1120 15:19:40.471731   283 init.cc:212]     @     0x5559b3297232 _PyEval_EvalFrameDefault
W1120 15:19:40.472141   283 init.cc:212]     @     0x5559b31de539 _PyEval_EvalCodeWithName
W1120 15:19:40.472515   283 init.cc:212]     @     0x5559b322df57 _PyFunction_FastCallKeywords
W1120 15:19:40.472954   283 init.cc:212]     @     0x5559b32968cc _PyEval_EvalFrameDefault
W1120 15:19:40.473354   283 init.cc:212]     @     0x5559b31de539 _PyEval_EvalCodeWithName
W1120 15:19:40.473727   283 init.cc:212]     @     0x5559b322df57 _PyFunction_FastCallKeywords
W1120 15:19:40.474166   283 init.cc:212]     @     0x5559b32968cc _PyEval_EvalFrameDefault
W1120 15:19:40.474566   283 init.cc:212]     @     0x5559b31de539 _PyEval_EvalCodeWithName
W1120 15:19:40.474992   283 init.cc:212]     @     0x5559b31df424 PyEval_EvalCodeEx
W1120 15:19:40.475389   283 init.cc:212]     @     0x5559b31df44c PyEval_EvalCode
W1120 15:19:40.475708   283 init.cc:212]     @     0x5559b32f4b74 run_mod
W1120 15:19:40.476099   283 init.cc:212]     @     0x5559b32feeb1 PyRun_FileExFlags
W1120 15:19:40.476526   283 init.cc:212]     @     0x5559b32ff0a3 PyRun_SimpleFileExFlags
W1120 15:19:40.476832   283 init.cc:212]     @     0x5559b3300195 pymain_main
Segmentation fault (core dumped)

**Comment by MRXLT** [comment]
你好，请参考以下页面中tensor转lod-tensor的代码
https://www.paddlepaddle.org.cn/documentation/docs/zh/user_guides/howto/basic_concept/lod_tensor.html#lod-tensor

**Comment by wwjjy** [comment]
@MRXLT 
https://github.com/PaddlePaddle/models/issues/3952#issuecomment-555876382 您好顺便请您帮我看一下，我这样的写法有哪里不对吗，报错的原因是什么呢？

**Comment by wwjjy** [comment]
同时，我也尝试了将数据转为lodtensor，报错如下：
![image](https://user-images.githubusercontent.com/20850734/69221163-9713de00-0bb1-11ea-80a0-91acecb3821d.png)
是我转换的类型不对吗》

**Comment by wanghaoshuang** [comment]
关于crnn产出的定长数据如何使用warpctc的问题，请参考warpctc的文档：
https://www.paddlepaddle.org.cn/documentation/docs/zh/api_cn/layers_cn/warpctc_cn.html#warpctc

另外您可以通过fluid.layers.Print查看label_length和pred_length的数值是否正确。

**Comment by wwjjy** [comment]
静态图版本代码如下：
![image](https://user-images.githubusercontent.com/20850734/69234698-bcf9ac80-0bca-11ea-94c1-f4455c37818d.png)
动态图版本代码如下：
![image](https://user-images.githubusercontent.com/20850734/69234719-ca169b80-0bca-11ea-9e89-8dbdf245cec2.png)
静态图可以成功运行，但是动态图的就会报段错误，数据都是采用一样的方式生成的，是我动态图哪里写的方式不对吗 @wanghaoshuang

**Comment by songyouwei** [comment]
这个问题的原因是文档中对label的解释不太清楚，没有明确input和label之间的关系，我们后面会对文档说明做一些补充和修正。

这里，label输入的shape改成`[batch_size, max_seq_length]`，确保`input_length`和`label_length`不为空，就没问题了：

```python
import paddle.fluid as fluid
import numpy as np
batch_size = 128
max_seq_length = 32
num_classes = 64
with fluid.dygraph.guard(fluid.CPUPlace()):
    x = fluid.dygraph.to_variable(np.random.rand(max_seq_length, batch_size, num_classes).astype('float32'))
    y = fluid.dygraph.to_variable(np.random.randint(0, num_classes-1, [batch_size, max_seq_length]).astype('int32'))
    x_len = fluid.dygraph.to_variable(np.array([max_seq_length] * batch_size).astype('int64'))
    y_len = fluid.dygraph.to_variable(np.array([max_seq_length] * batch_size).astype('int64'))
    cost = fluid.layers.warpctc(input=x, label=y, blank=num_classes-1, input_length=x_len, label_length=y_len)
```
